src/storage/txn/actions/common.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use tikv_kv::{Snapshot, SEEK_BOUND};
+use txn_types::{Key, TimeStamp, Write, WriteType};
+
+use crate::storage::mvcc::{Result, SnapshotReader};
+
+/// Returns the new `last_change_ts` and `versions_to_last_change` according
+/// to this write record. If it is unknown from the given write, try iterate to
+/// the last change and find the answer.
+pub fn next_last_change_info<S: Snapshot>(
+    key: &Key,
+    write: &Write,
+    start_ts: TimeStamp,
+    reader: &SnapshotReader<S>,
+    commit_ts: TimeStamp,
+) -> Result<(TimeStamp, u64)> {
+    match write.write_type {
+        WriteType::Put | WriteType::Delete => Ok((commit_ts, 1)),
+        WriteType::Lock | WriteType::Rollback => {
+            if !write.last_change_ts.is_zero() || write.versions_to_last_change != 0 {
```

Comment:
if the `last_change_ts` is not zero, it looks the `versions_to_last_change` should always be greater than zero, do we need to add a check and assertion here for the unexpected path?

Reply:
It is better to encapsulate the 2 variables into one enum so such worries can be avoided. I made an attempt but it seems to large for a bugfix. I will refactor it in a following PR.